src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -0,0 +1,158 @@
+
+  auto permute = Details::sortObjects(exec_space, weights);
+
+  Kokkos::Profiling::popRegion();
+
+  return permute;
+}
+
+template <typename Edges, typename UnionFind, typename Permute,
+          typename Parents>
+void dendrogramUnionFindHost(Edges edges_host, Permute permute,
+                             UnionFind union_find, Parents &parents_host)
+{
+  Kokkos::DefaultHostExecutionSpace exec_space;
+
+  int const num_edges = edges_host.size();
+  int const num_vertices = num_edges + 1;
+  auto const vertices_offset = num_edges;
+
+  constexpr int UNDEFINED = -1;
+
+  Kokkos::View<int *, Kokkos::HostSpace> set_edges_host(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::labels"),
+      num_vertices + num_edges);
+  Kokkos::deep_copy(exec_space, set_edges_host, UNDEFINED);
+
+  for (int k = 0; k < num_edges; ++k)
+  {
+    auto e = permute(k);
+
+    int i = edges_host(e).source;
+    int j = edges_host(e).target;
+
+    for (int s : {i, j})
+    {
+      auto child = set_edges_host(union_find.representative(s));
+      if (child == UNDEFINED)
+      {
+        // This will happen when a vertex has not been assigned a parent yet
+        parents_host(vertices_offset + s) = e;
+      }
+      else
+      {
+        parents_host(child) = e;
+      }
+    }
+
+    union_find.merge(i, j);
+
+    set_edges_host(union_find.representative(i)) = e;
+  }
+  parents_host(permute(num_edges - 1)) = UNDEFINED; // root
+}
+
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<int *, MemorySpace>
+dendrogramUnionFind(ExecutionSpace const &exec_space,
+                    Kokkos::View<WeightedEdge *, MemorySpace> edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::dendrogram_union_find");
+
+  Kokkos::Profiling::pushRegion(
+      "ArborX::Dendrogram::dendrogram_union_find::edge_sort");
+  auto permute = computeEdgesPermutation(exec_space, edges);
+  Kokkos::Profiling::popRegion();
+
+  auto const num_vertices = edges.size() + 1;
+
+  Kokkos::View<int *, MemorySpace> parents(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::edge_parents"),
+      2 * num_vertices - 1);
+
+  Kokkos::View<int *, MemorySpace> vertex_labels(
+      Kokkos::view_alloc(
+          Kokkos::WithoutInitializing,
+          "ArborX::Dendrogram::dendrogram_union_find::vertex_labels"),
+      num_vertices);
+  iota(exec_space, vertex_labels);
+
+  Kokkos::Profiling::pushRegion(
+      "ArborX::Dendrogram::dendrogram_union_find::copy_to_host");
+
+  auto edges_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, edges);
+  auto permute_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, permute);
+  auto parents_host = Kokkos::create_mirror_view(parents);
+  auto vertex_labels_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, vertex_labels);
```

Comment:
Is there any benefit on performing iota on the device and copy to the host rather than just create the view and initialize it directly on the host?